base/src/store.rs

```diff
@@ -0,0 +1,62 @@
+use crate::error::BlockError;
+use core::future::Future;
+use core::pin::Pin;
+use cid::Cid;
+use std::path::Path;
+
+pub type StoreResult<T> = Pin<Box<dyn Future<Output = Result<T, BlockError>> + Send>>;
+
+/// Visibility of a block.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Visibility {
+    /// Block is not announced on the network.
+    Private,
+    /// Block is only announced on the local network.
+    Local,
+    /// Block is announced.
+    Public,
+}
+
+/// Implementable by ipld storage backends.
+pub trait Store {
+    /// Returns a block from the store. If the block is not in the
+    /// store it fetches it from the network and pins the block. This
+    /// future should be wrapped in a timeout. Dropping the future
+    /// cancels the request.
+    fn get(&self, cid: &Cid) -> StoreResult<Box<[u8]>>;
+
+    /// Inserts and pins a block into the store and announces the block
+    /// if it is visible.
+    fn insert(&self, cid: &Cid, data: Box<[u8]>, visibility: Visibility) -> StoreResult<()>;
+
+    /// Flushes the write buffer.
+    fn flush(&self) -> StoreResult<()>;
+
+    /// Marks a block ready for garbage collection.
+    fn unpin(&self, cid: &Cid) -> StoreResult<()>;
+
+    /// Garbage collects unused blocks in the background.
+    fn gc(&self) -> StoreResult<()>;
+}
+
+/// Implemented by ipld storage backends that support multiple users.
+pub trait MultiUserStore: Store {
+    /// Pin a block.
+    ///
+    /// This creates a symlink chain from root -> path -> block. The block is unpinned by
+    /// breaking the symlink chain.
+    fn pin(&self, cid: &Cid, path: &Path) -> StoreResult<()>;
```

Comment:
I think i see why this is here, though having `pin` and `unpin` split across two separate traits makes me a little dizzy. :sweat_smile: What other functions would be added to MultiUserStore in the future.

Would the different types of stores be more suited as an enum? Basically `rust-ipld`'s way of broadcasting outward: "these are the different types of stores I know what to do with"

Reply:
In a single user store I'm not sure it makes sense to `pin` a block. Blocks should be pinned on insert and unpinned when they are no longer needed. Unpin is similar to a remove in this case, except that because the store is content addressed, proper liveness tracking needs to happen.

In a multi user store it makes sense. The idea is that you have something like this:
```
user            path
ipfs             /ipfs/block/{cid} 
ipfs             /ipfs/roots/{hash(path)} -> /home/aphelionz/.config/application/{cid}
aphelionz  /home/aphelionz/.config/orbitdb/{cid} -> /ipfs/block/{cid}
```

Reply:
I'm cool with this as it saves the implementer the trouble of having to include pin if they don't need it